src/Microsoft.Azure.WebJobs.Host/Executors/JobHostContextFactory.cs

```diff
@@ -27,121 +27,162 @@
 
 namespace Microsoft.Azure.WebJobs.Host.Executors
 {
-    internal class JobHostContextFactory : IJobHostContextFactory
+    internal static class JobHostConfigurationExtensions
     {
-        private readonly IStorageAccountProvider _storageAccountProvider;
-        private readonly IConsoleProvider _consoleProvider;
-        private readonly JobHostConfiguration _config;
-
-        public JobHostContextFactory(IStorageAccountProvider storageAccountProvider, IConsoleProvider consoleProvider, JobHostConfiguration config)
+        // Do full initialization (both static and runtime). 
+        // This can be called multiple times on a config. 
+        public static async Task<JobHostContext> CreateAndLogHostStartedAsync(
+            this JobHostConfiguration config, 
+            JobHost host, 
+            CancellationToken shutdownToken, 
+            CancellationToken cancellationToken)
         {
-            _storageAccountProvider = storageAccountProvider;
-            _consoleProvider = consoleProvider;
-            _config = config;
+            var newServices = config.DoStaticInitialization();
+            JobHostContext context = await newServices.DoRuntimeInitialization(config, host, shutdownToken, cancellationToken);
```

Comment:
In line with the other renames I suggested for clarity, I'd rename stuff so you can do:

var context = await CreateJobHostContextAsync(jobHostServices, host, ...);